src/catalog/src/information_schema/predicate.rs

```diff
@@ -0,0 +1,622 @@
+                                return None;
+                            };
+
+                            values.push(value);
+                        }
+
+                        let predicate = Predicate::InList(c.name, values);
+
+                        if negated {
+                            Some(Predicate::Not(Box::new(predicate)))
+                        } else {
+                            Some(predicate)
+                        }
+                    }
+                    _ => None,
+                }
+            }
+            _ => None,
+        }
+    }
+}
+
+/// Perform SQL left LIKE right, return `None` if fail to evaluate.
+/// - `s` the target string
+/// - `pattern` the pattern just like '%abc'
+/// - `case_insensitive` whether to perform case-insensitive like or not.
+fn like_utf8(s: &str, pattern: &str, case_insensitive: &bool) -> Option<bool> {
+    let array = StringArray::from(vec![s]);
+    let patterns = StringArray::new_scalar(pattern);
+
+    let Ok(booleans) = (if *case_insensitive {
+        comparison::ilike(&array, &patterns)
+    } else {
+        comparison::like(&array, &patterns)
+    }) else {
+        return None;
+    };
+
+    // Safety: at least one value in result
+    Some(booleans.value(0))
+}
+
+fn is_string_literal(expr: &DfExpr) -> bool {
+    matches!(expr, DfExpr::Literal(ScalarValue::Utf8(Some(_))))
+}
+
+fn is_column(expr: &DfExpr) -> bool {
+    matches!(expr, DfExpr::Column(_))
+}
+
+/// A list of predicate
+pub struct Predicates {
+    predicates: Vec<Predicate>,
+}
+
+impl Predicates {
+    /// Try its best to create predicates from [`ScanRequest`].
+    pub fn from_scan_request(request: &Option<ScanRequest>) -> Predicates {
+        if let Some(request) = request {
+            let mut predicates = Vec::with_capacity(request.filters.len());
+
+            for filter in &request.filters {
+                if let Some(predicate) = Predicate::from_expr(filter.df_expr().clone()) {
+                    predicates.push(predicate);
+                }
+            }
+
+            Self { predicates }
+        } else {
+            Self {
+                predicates: Vec::new(),
+            }
+        }
+    }
+
+    /// Evaluate the predicates with the row.
+    /// returns true when all the predicates are satisfied or can't be evaluated.
+    pub fn eval(&self, row: &[(&str, &Value)]) -> bool {
+        // fast path
+        if self.predicates.is_empty() {
+            return true;
+        }
+
+        let mut result = true;
+
+        for predicate in &self.predicates {
+            match predicate.eval(row) {
+                Some(b) => {
+                    result = result && b;
+                }
+                // The predicate can't evaluate with the row, continue
+                None => continue,
+            }
+
+            if !result {
+                break;
+            }
+        }
+
+        result
```

Comment:
Is it possible to rewrite it like this?
```suggestion
        self.predicates.iter().map(|p| p.eval(row)).flatten().all(|x| x)
```

Reply:
Looks like we can't, because the result of `p.eval(row)` is an `Option`.

Reply:
`flatten` can skip `None`s, and `all` can return early as well.

Reply:
> `flatten` can skip `None`s, and `all` can return early as well.

Let me give a try

Reply:
It works!

```rust
        self.predicates
            .iter()
            .filter_map(|p| p.eval(row))
            .all(|b| b)
```

Reply:
a shorter way:

```rust
self.predicates.iter().filter_map(|p| p.eval(row)).all(|x| x)
```